language/evm/move-to-yul/src/generator.rs

```diff
@@ -0,0 +1,770 @@
+            .collect_vec();
+        if creators.len() > 1 {
+            ctx.env
+                .error(&creators[1].get_loc(), "multiple #[create] functions")
+        }
+        if let Some(creator) = creators.pop() {
+            ctx.check_no_generics(&creator);
+            // TODO: implement creators
+            ctx.env
+                .error(&creators[1].get_loc(), "[NYI] #[create] functions")
+        }
+    }
+
+    /// Generate Yul definitions for all callable functions.
+    fn callable_functions(&mut self, ctx: &Context, module: &ModuleEnv<'_>) {
+        let callables = module
+            .get_functions()
+            .filter(|f| is_callable_fun(f))
+            .collect_vec();
+        // TODO: generate dispatcher and marshalling
+        for callable in callables {
+            ctx.check_no_generics(&callable);
+            self.function(
+                ctx,
+                &callable
+                    .module_env
+                    .get_id()
+                    .qualified(callable.get_id())
+                    .instantiate(vec![]),
+            )
+        }
+    }
+
+    /// Begin a new code block.
+    fn begin_code_block(&mut self, ctx: &Context) {
+        assert!(self.needed_move_functions.is_empty());
+        assert!(self.needed_yul_functions.is_empty());
+        emitln!(ctx.writer, "code {");
+        ctx.writer.indent();
+    }
+
+    /// End a code block, generating all functions needed by top-level callable functions.
+    fn end_code_block(&mut self, ctx: &Context) {
+        // Before the end of the code block, we need to emit definitions of all
+        // functions reached by callable entry points. While we traversing this list,
+        // more functions might be added due to transitive calls.
+        while let Some(fun_id) = self.needed_move_functions.pop() {
+            self.done_move_functions.insert(fun_id.clone());
+            self.function(ctx, &fun_id)
+        }
+
+        // We finally emit code for all Yul functions which have been needed by the Move
+        // functions.
+        for fun in &self.needed_yul_functions {
+            emitln!(ctx.writer, &fun.yule_def());
+        }
+        ctx.writer.unindent();
+        emitln!(ctx.writer, "}")
+    }
+}
+
+// ================================================================================================
+// Function generation
+
+impl Generator {
+    /// Generate Yul function for Move function.
+    fn function(&mut self, ctx: &Context, fun_id: &QualifiedInstId<FunId>) {
+        let fun = &ctx.env.get_function(fun_id.to_qualified_id());
+        // Emit function header
+        let params = (0..fun.get_parameter_count())
+            .map(|idx| ctx.make_local_name(fun, idx))
+            .join(", ");
+        let results = match fun.get_return_count() {
+            0 => "".to_string(),
+            1 => format!(" -> {}", ctx.make_result_name(fun, 0)),
+            _ => format!(
+                " -> ({})",
```

Comment:
In Yul, "()" is not needed for returning multiple values

Reply:
Done